### ISSUE TITLE
Ambient: leaked pod in snapshot, leading to ztunnel failing to startup

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#53371]
Scenario: left a cluster running for a few weeks. Not sure everything happening during this time. CNI is version 455eb1d46e.

When I come back, I try to restart ztunnel. It fails to start one of the workloads sent in a snapshot. Upon debugging, I find the pod its trying to start no longer exists.

Ztunnel logs (the old one):
```
2024-09-24T21:47:50.474706Z     info    inpod::statemanager     pod received, starting proxy    uid="f57c48e4-77af-4892-b48c-0e318cbcb8f3" name="shell-6d8bcd654d-bshz9" namespace="default"
2024-09-26T15:20:55.119650Z     info    drain:proxy{wl=default/shell-6d8bcd654d-bshz9}  shutdown complete       component="inbound"
2024-09-26T15:20:55.120259Z     info    dns::server:dns_proxy{wl=default/shell-6d8bcd654d-bshz9}        shutting down the DNS server
2024-09-26T15:20:55.120291Z     info    dns::server:dns_proxy{wl=default/shell-6d8bcd654d-bshz9}        dns server drained
2024-09-26T15:20:55.120639Z     info    drain:proxy{wl=default/shell-6d8bcd654d-bshz9}  shutdown complete       component="outbound"
2024-09-26T15:20:55.121417Z     info    drain:proxy{wl=default/shell-6d8bcd654d-bshz9}  shutdown complete       component="inbound passthrough"
```

So here we see it start, then 2 days later shutdown.

CNI logs
```
2024-09-17T18:42:48.307477Z     info    cni-agent       sending pod to ztunnel as part of snapshot      uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
2024-09-17T18:42:51.790028Z     info    cni-agent       sending pod to ztunnel as part of snapshot      uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
2024-09-19T20:15:35.848966Z     info    cni-agent       sending pod to ztunnel as part of snapshot      uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
2024-09-24T21:47:50.474404Z     info    cni-agent       sending pod to ztunnel as part of snapshot      uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
2024-09-26T15:21:00.248303Z     info    cni-agent       adding pod to the mesh  ns=default name=shell-6d8bcd654d-bshz9
COMMIT  ns=default name=shell-6d8bcd654d-bshz9
COMMIT  ns=default name=shell-6d8bcd654d-bshz9
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
2024-09-26T15:21:00.312084Z     error   cni-agent       failed to update POD inpod: default/shell-6d8bcd654d-bshz9 exit status 1
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
2024-09-26T18:57:02.078870Z     info    cni-agent       adding pod to the mesh  ns=default name=shell-6d8bcd654d-bshz9
COMMIT  ns=default name=shell-6d8bcd654d-bshz9
COMMIT  ns=default name=shell-6d8bcd654d-bshz9
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
2024-09-26T18:57:02.121669Z     error   cni-agent       failed to update POD inpod: default/shell-6d8bcd654d-bshz9 exit status 1
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
exit status 1   ns=default name=shell-6d8bcd654d-bshz9
2024-10-01T14:46:42.442665Z     info    cni-agent       sending pod to ztunnel as part of snapshot      uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
2024-10-01T14:46:42.457361Z     error   cni-agent       add-workload: got ack error: failure to start proxy : Bind([::1]:15053, Os { code: 99, kind: AddrNotAvailable, message: "Cannot assign requested address" })  uid=f57c48e4-77af-4892-b48c-0e318cbcb8f3 name=shell-6d8bcd654d-bshz9 namespace=default serviceAccount=default
```

Here we see it sends as a snapshot a few times (presumably the old ztunnel was not running as long as the CNI, hence we miss a few of the original snapshots).

While Ztunnel sees a delete at 09-26T15:20:55, CNI sees an **add** 5s later (09-26T15:21:00). CNI doesn't log 'delete' at info level so presumably it got a delete. 

Its unclear what happened in kubernetes to the Pod unfortunately.. all I know is that it did not exist on 10/1.

**Comment by bleggett** [comment]
> For logging perspective, making 'delete' info level and maybe logging the source of operations (from informer or from CNI plugin) would be useful.

The only operation the plugin can initiate is `ADD` ATM - making the origin of the ADD clearer in logs would be good.

The informer cache should never have pods that don't exist in it, that's odd.

**Comment by howardjohn** [comment]
> The informer cache should never have pods that don't exist in it, that's odd.

I doubt its the informer cache itself but our snapshot of the netns which is our cache-of-a-cache.

**Comment by bleggett** [comment]
> > The informer cache should never have pods that don't exist in it, that's odd.
> 
> I doubt its the informer cache itself but our snapshot of the netns which is our cache-of-a-cache.

That's what I meant - yeah - the informer's pod snapshot cache (not the actual/generic k8s informer cache).

**Comment by bleggett** [comment]
I suspect this might be logically similar to the ipset issues we had - this might happen if `Add` fails.

We `ensure` the pod UID is in the snapshot as a first concern on Add, and only remove it on an explicit informer pod Remove.

**Comment by bleggett** [comment]
I think we need to actually revisit the concept of `PartialAdd`

1. We should (re)construct the snapshot on every ztunnel connect, rather than maintain it long term as mutable state.
2. PodAdd should just unconditionally require an active ztunnel connection, or fail/block, and doesn't need to mutate the snapshot state.


-

**Comment by keithmattix** [comment]
> We ensure the pod UID is in the snapshot as a first concern on Add, and only remove it on an explicit informer pod Remove - an Add failure that never triggers a Remove for instance would leave a stale pod UID in the informer pod cache.

Yeah this is definitely a bug; we sure ensure that anything that gets added preemptively gets removed on failure. 

> PodAdd should just unconditionally require an active ztunnel connection, or fail/block, and doesn't need to mutate the snapshot state.

Are you referring explicitly to the CNI ADD codepath? Might get tricky if we block here. Assuming that ztunnel always retries seems fine; if I'm understanding everything correctly, the core issue is that the istio-cni snapshot generation always assumes a happy path

**Comment by bleggett** [comment]
> Are you referring explicitly to the CNI ADD codepath? Might get tricky if we block here. 
> Assuming that ztunnel always retries seems fine

At some point it should probably stop retrying, but yeah by itself that's fine.

> if I'm understanding everything correctly, the core issue is that the istio-cni snapshot generation always assumes a happy path

Specifically: `cni/pkg/nodeagent/net.go` -> `AddPodToMesh`

The problem is that snapshot + ztunnel infinite retries is also used as an implicit queue/buffer for adds so that we don't have to block.

To keep that buffering we have to proactively add, but that means we don't have a reconciliation window to reactively remove (outside of kube deletion events) because it's a fire-and-forget thing.

It's fixable but it's not just a matter of happy path. Handling retries from the CNI agent is probably going to work better.

**Comment by bleggett** [comment]
xref of a (likely related) repo in Slack: https://istio.slack.com/archives/C37A4KAAD/p1731496728267749

2 problems that IMO both need to be fixed:

- ztunnel can't tolerate a snapshot with an invalid/missing entry

- `istio-cni` somehow sends a snapshot with one

**Comment by ilrudie** [comment]
Removed assignment to Ben. I don't think he's still working on this anymore 😔

**Comment by ilrudie** [comment]
https://github.com/istio/istio/pull/56738 fixes a pod leaked into the snapshot
